cmd/metal-api/internal/service/switch-service.go

```diff
@@ -244,9 +256,107 @@ func (r *switchResource) notifySwitch(request *restful.Request, response *restfu
 		return
 	}
 
+	oldSwitch, err := r.ds.FindSwitch(id)
+	if err != nil {
+		r.sendError(request, response, defaultError(err))
+		return
+	}
+
+	newSwitch := *oldSwitch
+	switchUpdated := false
+	for i, nic := range newSwitch.Nics {
+		state, has := requestPayload.PortStates[strings.ToLower(nic.Name)]
+		if has {
+			reported := metal.SwitchPortStatus(state)
+			newstate, changed := nic.State.SetState(reported)
+			if changed {
+				newSwitch.Nics[i].State = &newstate
+				switchUpdated = true
+			}
+		} else {
+			// this should NEVER happen; if the switch reports the state of an unknown port
+			// we log this and ignore it, but something is REALLY wrong in this case
+			r.log.Error("unknown switch port", "id", id, "nic", nic.Name)
+		}
+	}
+
+	if switchUpdated {
+		if err := r.ds.UpdateSwitch(oldSwitch, &newSwitch); err != nil {
+			r.sendError(request, response, defaultError(err))
+			return
+		}
+	}
+
 	r.send(request, response, http.StatusOK, v1.NewSwitchNotifyResponse(&newSS))
 }
 
+// toggleSwitchPort handles a request to toggle the state of a port on a switch. It reads the request body, validates the requested status is concrete, finds the switch, updates its NIC state if needed, and returns the updated switch on success.
+// toggleSwitchPort handles a request to toggle the state of a port on a switch. It reads the request body to get the switch ID, NIC name and desired state. It finds the switch, updates the state of the matching NIC if needed, and returns the updated switch on success.
+func (r *switchResource) toggleSwitchPort(request *restful.Request, response *restful.Response) {
+	var requestPayload v1.SwitchPortToggleRequest
+	err := request.ReadEntity(&requestPayload)
+	if err != nil {
+		r.sendError(request, response, httperrors.BadRequest(err))
+		return
+	}
+
+	desired := metal.SwitchPortStatus(requestPayload.Status)
+
+	if !desired.IsConcrete() {
+		r.sendError(request, response, httperrors.BadRequest(fmt.Errorf("the status %q must be concrete", requestPayload.Status)))
+		return
+	}
+
+	id := request.PathParameter("id")
+	oldSwitch, err := r.ds.FindSwitch(id)
+	if err != nil {
+		r.sendError(request, response, defaultError(err))
+		return
+	}
+
+	newSwitch := *oldSwitch
+	updated := false
+	found := false
+
+	// Updates the state of a NIC on the switch if the requested state change is valid
+	//
+	// Loops through each NIC on the switch and checks if the name matches the
+	// requested NIC. If a match is found, it tries to update the NIC's state to
+	// the requested state. If the state change is valid, it sets the new state on
+	// the NIC and marks that an update was made.
+	for i, nic := range newSwitch.Nics {
+		// compare nic-names case-insensitive
+		if strings.EqualFold(nic.Name, requestPayload.NicName) {
+			found = true
+			newstate, changed := nic.State.WantState(desired)
+			if changed {
+				newSwitch.Nics[i].State = &newstate
+				updated = true
+				break
```

Comment:
Could be placed after the if-condition because the iteration can always stop as soon as the nic was found.